pkg/datapath/linux/ipsec/xfrm_collector.go

```diff
@@ -0,0 +1,277 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2022 Authors of Cilium
+
+package ipsec
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/procfs"
+
+	"github.com/cilium/cilium/pkg/metrics"
+)
+
+const (
+	labelErrorType         = "type"
+	labelErrorTypeInbound  = "inbound"
+	labelErrorTypeOutbound = "inbound"
+)
+
+type xfrmCollector struct {
+	xfrmStatFunc func() (procfs.XfrmStat, error)
+
+	// Inbound errors
+	xfrmInErrorDesc           *prometheus.Desc
+	xfrmInBufferErrorDesc     *prometheus.Desc
+	xfrmInHdrErrorDesc        *prometheus.Desc
+	xfrmInNoStatesDesc        *prometheus.Desc
+	xfrmInStateProtoErrorDesc *prometheus.Desc
+	xfrmInStateModeErrorDesc  *prometheus.Desc
+	xfrmInStateSeqErrorDesc   *prometheus.Desc
+	xfrmInStateExpiredDesc    *prometheus.Desc
+	xfrmInStateMismatchDesc   *prometheus.Desc
+	xfrmInStateInvalidDesc    *prometheus.Desc
+	xfrmInTmplMismatchDesc    *prometheus.Desc
+	xfrmInNoPolsDesc          *prometheus.Desc
+	xfrmInPolBlockDesc        *prometheus.Desc
+	xfrmInPolErrorDesc        *prometheus.Desc
+	xfrmAcquireErrorDesc      *prometheus.Desc
+	xfrmFwdHdrErrorDesc       *prometheus.Desc
+
+	// Outbound errors
+	xfrmOutErrorDesc            *prometheus.Desc
+	xfrmOutBundleGenErrorDesc   *prometheus.Desc
+	xfrmOutBundleCheckErrorDesc *prometheus.Desc
+	xfrmOutNoStatesDesc         *prometheus.Desc
+	xfrmOutStateProtoErrorDesc  *prometheus.Desc
+	xfrmOutStateModeErrorDesc   *prometheus.Desc
+	xfrmOutStateSeqErrorDesc    *prometheus.Desc
+	xfrmOutStateExpiredDesc     *prometheus.Desc
+	xfrmOutPolBlockDesc         *prometheus.Desc
+	xfrmOutPolDeadDesc          *prometheus.Desc
+	xfrmOutPolErrorDesc         *prometheus.Desc
+	xfrmOutStateInvalidDesc     *prometheus.Desc
```

Comment:
These are a lot of metrics :exploding_head: Was it considered having `In` and `Out` and all the error types as labels of one single metric instead?

Reply:
making sense to me, I have collated all into one single metric cilium_ipsec_xfrm_error{error=<reason>,type=inbound|outbound}